src/group_theory/submonoid/membership.lean

```diff
@@ -206,12 +206,76 @@ set.ext (λ n, exists_congr $ λ i, by simp; refl)
 
 @[simp] lemma mem_powers (n : M) : n ∈ powers n := ⟨1, pow_one _⟩
 
+lemma mem_powers_iff (z x : M) : x ∈ powers z ↔ ∃ n : ℕ, z ^ n = x := iff.rfl
+
 lemma powers_eq_closure (n : M) : powers n = closure {n} :=
 by { ext, exact mem_closure_singleton.symm }
 
 lemma powers_subset {n : M} {P : submonoid M} (h : n ∈ P) : powers n ≤ P :=
 λ x hx, match x, hx with _, ⟨i, rfl⟩ := P.pow_mem h i end
 
+/-- Exponentiation map from natural numbers to powers. -/
+def pow (n : M) (m : ℕ) : powers n := ⟨n ^ m, m, rfl⟩
+
+/-- Logarithms from powers to natural numbers. -/
+def log [decidable_eq M] {n : M} (p : powers n) : ℕ :=
+nat.find $ (mem_powers_iff n p).mp p.prop
+
+@[simp] theorem pow_log_eq_self [decidable_eq M] {n : M} (p : powers n) : pow n (log p) = p :=
+begin
+  rw [pow, log],
+  rcases p with ⟨_, hp⟩,
+  congr,
+  exact nat.find_spec hp,
+end
+
+lemma pow_right_injective_iff_pow_injective {n : M} :
+  function.injective (λ m : ℕ, n ^ m) ↔ function.injective (pow n) :=
+begin
+  split; simp only []; intros h₁ _ _ h₂,
+  { simp only [pow, subtype.mk_eq_mk] at h₂,
+    exact h₁ h₂ },
+  { dsimp only [function.injective, pow] at h₁,
+    apply h₁,
+    congr,
+    exact h₂ }
+end
+
+theorem log_pow_eq_self [decidable_eq M] {n : M} (h : function.injective (λ m : ℕ, n ^ m)) (m : ℕ) :
+  log (pow n m) = m := pow_right_injective_iff_pow_injective.mp h $ pow_log_eq_self _
+
+-- TODO: Move this to data.int.basic
+lemma int.pow_right_injective {x : ℤ} (h : 2 ≤ x.nat_abs) : function.injective (λ (n : ℕ), x ^ n) :=
+λ n m hnm, begin
+  obtain pos | neg := int.nat_abs_eq x,
+  { lift x to ℕ using (pos.symm ▸ int.coe_nat_nonneg _),
+  norm_cast at h hnm, -- norm_cast not simplifying enough in data.int.basic
+  exact nat.pow_right_injective h hnm },
+  dsimp only at hnm,
+  rw [←neg_neg x, neg_pow, neg_pow (-x)] at hnm,
+  replace neg := neg_eq_iff_neg_eq.mp neg.symm,
+  generalize' hy : -x = y,
+  rw [←neg_neg x, int.nat_abs_neg] at h,
+  rw [hy] at *,
+  lift y to ℕ using (neg.symm ▸ int.coe_nat_nonneg _),
+  obtain h₁ | h₁ := neg_one_pow_eq_or ℤ n;
+  obtain h₂ | h₂ := neg_one_pow_eq_or ℤ m;
+  all_goals { simp only [h₁, h₂, neg_mul_eq_neg_mul_symm, one_mul, neg_inj] at hnm,
+              norm_cast at h hnm,
+              exact nat.pow_right_injective h hnm <|> exfalso },
+  swap,
+  replace hnm := hnm.symm,
+  all_goals { apply int.ne_neg_of_pos _ _ hnm;
+              norm_cast;
+              apply nat.pos_of_ne_zero;
+              apply pow_ne_zero;
+              apply ne_bot_of_gt;
+              exact nat.succ_le_iff.mp h }
+end
```

Comment:
Can you split out a small PR that puts this lemma in the right place?

Reply:
Opened PR #8278